---
title: "Processing: Quality Control"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Processing: Quality Control}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

 **gwasglue2**  deals with some inconsistencies in GWAS summary data, such as variants with the same chromosomal position and alleles that have different betas, p-values and allele frequencies. This is a common problem in GWAS summary data, and **gwasglue2** removes this problematic variants.

<br>
<br>

## Setup
In addition to **gwasglue2**, weâ€™ll load the following package:

- [**ieugwasr**](https://mrcieu.github.io/ieugwasr/) - R package for accessing and query the IEU GWAS database

```{r setup}
library(gwasglue2)
library(ieugwasr)
```

<br>
<br>

## Functions
* `create_summaryset()` using the `qc = TRUE` parameter to perform quality control on the GWAS summary data.

<br>
<br>




## Example
We are using as example GWAS summary data obtained from IEU OpenGWAS through the **ieugwasr** package (see [here](input-opengwas.html) for details). The data belongs to the `finn-b-I9_CHD` study of cardiac heart disease (chd), in the HMGCR gene region (chr5:74132993-75132993). 

First, we retrieve the GWAS summary data and then create the `SummarySet`.

```{r}
 data_finn <- ieugwasr::associations(variants = "5:74132993-75132993", id = "finn-b-I9_CHD")

 sumset3 <- create_summaryset(data_finn)
```

There is a warning creating the `SummarySet`, telling us that there is a problem with the GWAS summary data. Specifically, there are variants with the same chromosomal position and alleles that have different betas, p-values and allele frequencies. If we change the quality control parameter to `qc = TRUE`, we will let **gwasglue2**  deal with these inconsistencies in the data.

```{r}
sumset3 <- create_summaryset(data_finn, qc = TRUE)
```