---
title: "Tutorial2"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Tutorial2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```



#### This tutorial is still work in progress.



## 1. Alternative regional genotype-phenotype map (WIP)


```{r  eval= FALSE}
remotes::install_github('ritarasteiro/hyprcoloc', build_opts = c('--resave-data', '--no-manual'), build_vignettes = TRUE) 
remotes::install_github('ritarasteiro/susieR')

```

```{r setup}
#library(gwasglue2)
devtools::load_all("../") # this was added just for development
library(susieR)     # fork that takes Summaryset
library(hyprcoloc)  # fork that takes Dataset

```

### HMGCR gene
We'll analyse the following traits (chd - cardiac heart disease; ldl - low-density lipoprotein; hdl - high-density lipoprotein; trig - triglycerides and t2d - type 2 diabetes) around the HMGCR (3-hydroxy-3-methylglutaryl-CoA reductase) gene region. 

First we will set the path to where the reference population plink files are located to build the LD matrix.

```{r}
#ld_ref <- "/project/data/reference/ld/EUR"
ld_ref <- "../data/ld/EUR"
```

Then we choose the data for each trait and set the IEU IDs. 



```{r}
chd_id <- "ieu-a-7"
ldl_id <- "ieu-b-110"
hdl_id <- "ieu-b-109"
trig_id <- "ieu-b-111"
t2d_id <- "ebi-a-GCST006867"
ids <- c(chd_id, ldl_id, hdl_id, trig_id, t2d_id)
```



Create a gwasglue 2 DataSet object  and harmonise trait against trait, trait against LD matrix

```{r}

# Get GWAS summary statistics
hmgcr_chrpos <- "5:74132993-75132993"
hmgcr_data <- lapply(seq_along(ids), function(i){
    s <- ieugwasr::associations(variants = hmgcr_chrpos, id = ids[i])
})


# create gwasglue2 DataSet object (hmgcr)
hmgcr <- createDataSet(hmgcr_data,ids=ids, tools = c("finemap","coloc"), source = "IEUopenGWAS", harmonise = TRUE, tolerance = 0.08, action = 1,strand= "forward") %>% 
  harmonise_ld(., ld_ref = ld_ref, bfile = TRUE, plink_bin = "plink") 
```

Before starting with the fine mapping and colocalisation analyses, we are going to look at the p-values for each variant in each trait. 

```{r fig.dim = c(8, 12)}
# plot gwasglue2 DataSet object (hmgcr)

plot_gwasglue(hmgcr, type="manhattan", title = "HMGCR")

```

After summary sets are harmonised, marginalise each summary set independently and create a new dataset with all marginalised summary sets merged

```{r}
# do finemapping with susie
ntraits <- getLength(hmgcr)
hmgcr_marginalised <- lapply(1:ntraits, function(trait)
  {
    # Takes in 1 SS
    # Outputs 1 DS (with at least 1 SS)
    ds <- susie_rss(R = getLDMatrix(hmgcr), summaryset = getSummarySet(hmgcr, trait))
})
```

 Now we are going to merge the datasets. Note that the `merge_datasets()` function is just going to merge the marginalised datasets. 

```{r}
hmgcr_marginalised <- merge_datasets(hmgcr_marginalised)

``` 

 Try hyprcoloc with raw datasets. (Use fork of hypercoloc that allows a dataset object to be provided)

```{r} 

res_hmgcr <- hyprcoloc(dataset = hmgcr) 
#hyprcoloc is using gwasglue2 DataSet class object as input
print(res_hmgcr)

```

Try hyprcoloc with marginalised datasets

```{r}

res_hmgcr_marginalised <- hyprcoloc(dataset = hmgcr_marginalised)
print(res_hmgcr_marginalised)
```

Also try with these other regions:


```{r}
pcsk9_chrpos <- "1:55005221-56005221"
npc1l1_chrpos <- "7:44052134-45052134"
lpa_chrpos <- "6:160952515-161087407"
```


### PCSK9 (proprotein convertase subtilisin/kexin type 9) gene region

Create dataset and harmonise trait against trait, trait against LD matrix (note that we are not saving the SummarySet objects separately, but filling them directly to the DataSet).

```{r eval= FALSE}
# create gwasglue2 DataSet object (hmgcr)
pcsk9_chrpos <- "1:55005221-56005221"

chd_dat <- createDataSet(traits=c("ieu-a-7", "ukb-d-I9_IHD"), variants = pcsk9_chrpos, tools = c("finemap","coloc"), source = "IEUopenGWAS", harmonise = TRUE, tolerance = 0.08, action = 1) %>% 
  harmonise_ld(., ld_ref = ld_ref, bfile = TRUE, plink_bin = "plink") 
 
chd_meta <- meta_analyse(chd_dat, method="fixed")
```

```{r}
# create gwasglue2 DataSet object (pcsk9)
pcsk9 <- createDataSet(traits = ids, variants = pcsk9_chrpos, tools = c("finemap","coloc"), source = "IEUopenGWAS", harmonise = TRUE, tolerance = 0.08, action = 1) %>% 
  harmonise_ld(., ld_ref = ld_ref, bfile = TRUE, plink_bin = "plink") 

```

Manhattan plots for each trait in the DataSet.

```{r fig.dim = c(8, 12)}
# plot gwasglue2 DataSet object (pcsk9)

plot_gwasglue(pcsk9, type="manhattan", title = "PCSK9")


```
Marginalise each summary set independently and create a new dataset with all marginalised summary sets merged

```{r}
# do finemapping with susie 
ntraits <- getLength(pcsk9)
pcsk9_marginalised <- lapply(1:ntraits, function(trait)
  {
    # Takes in 1 SS
    # Outputs 1 DS (with at least 1 SS)
    ds <- susie_rss(R = getLDMatrix(pcsk9), summaryset = getSummarySet(pcsk9, trait))
 }) 

```

Merge the marginalised datasets:

```{r}
pcsk9_marginalised <- merge_datasets(pcsk9_marginalised)
```


Try hyprcoloc with raw datasets. 

```{r}
res_pcsk9 <- hyprcoloc(dataset = pcsk9)
print(res_pcsk9)

```

Try hyprcoloc with marginalised datasets

```{r}
res_pcsk9_marginalised <- hyprcoloc(dataset =pcsk9_marginalised)
print(res_pcsk9_marginalised)

```

### NPC1L1 (NPC1 like intracellular cholesterol transporter 1) gene region
Create dataset and harmonise trait against trait, trait against LD matrix

```{r}
# create gwasglue2 DataSet object (npc1l1)
npc1l1 <- createDataSet(traits=ids, variants = npc1l1_chrpos, tools = c("finemap","coloc"), source = "IEUopenGWAS", harmonise = TRUE, tolerance = 0.08, action = 1) %>% 
  harmonise_ld(., ld_ref = ld_ref, bfile = TRUE, plink_bin = "plink") 
```


Manhattan plots for each trait in the DataSet.

```{r fig.dim = c(8, 12)}
# plot gwasglue2 DataSet object (npc1l1)

plot_gwasglue(npc1l1, type="manhattan", title ="NPC1L1")
```

Marginalise each summary set independently and create a new dataset with all marginalised summary sets merged 

```{r}
ntraits <- getLength( npc1l1)
npc1l1_marginalised <- lapply(1:ntraits, function(trait)
  {
    # Takes in 1 SS
    # Outputs 1 DS (with at least 1 SS)
    ds <- susie_rss(R = getLDMatrix(npc1l1), summaryset = getSummarySet( npc1l1, trait))
}) 
```

Merge the marginalised datasets:

```{r}
npc1l1_marginalised <- merge_datasets(npc1l1_marginalised)
```


Try hyprcoloc with raw datasets. 

```{r}
res_npc1l1 <- hyprcoloc(dataset = npc1l1)
print(res_npc1l1)

```

Try hyprcoloc with marginalised datasets

```{r}
res_npc1l1_marginalised <- hyprcoloc(dataset =npc1l1_marginalised)
print(res_npc1l1_marginalised)

```

### LPA (Lipoprotein(A)) gene region
Create dataset and harmonise trait against trait, trait against LD matrix

```{r}
# create gwasglue2 DataSet object (lpa)
lpa <- createDataSet(traits=ids, variants = lpa_chrpos, tools = c("finemap","coloc"), source = "IEUopenGWAS", harmonise = TRUE, tolerance = 0.08, action = 1) %>% 
  harmonise_ld(., ld_ref = ld_ref, bfile = TRUE, plink_bin = "plink") 

```

Manhattan plots for each trait in the DataSet.

```{r fig.dim = c(8, 12)}
# plot gwasglue2 DataSet object (lpa)

 plot_gwasglue(lpa, type="manhattan", title = "LPA")

```

Marginalise each summary set independently and create a new dataset with all marginalised summary sets merged

```{r}
ntraits <- getLength(lpa)
lpa_marginalised <- lapply(1:ntraits, function(trait)
{
  # Takes in 1 SS
  # Outputs 1 DS (with at least 1 SS)
  ds <- susie_rss(R = getLDMatrix(lpa), summaryset = getSummarySet(lpa, trait))
}) 

```

Merge the marginalised datasets:

```{r}
lpa_marginalised <- merge_datasets(lpa_marginalised)

```


Try hyprcoloc with raw datasets. 

```{r}

res_lpa <- hyprcoloc(dataset = lpa)
print(res_lpa)

```

Try hyprcoloc with marginalised datasets

```{r}
res_lpa_marginalised <- hyprcoloc(dataset = lpa_marginalised)
print(res_lpa_marginalised)
```


