---
title: "Tutorial"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

```{r setup}
library(gwasglue2)
library(TwoSampleMR)
library(ieugwasr)
library(hyprcoloc)
library(susieR)
```

There are two usecases for this tutorial 

1. Performing a basic TwoSampleMR analysis
2. Generating a regional genotype-phenotype map (Million Veteran's Programme analysis example)


#### This tutorial is still work in progress. It reflects an earlier draft of gwasglue2.

## 1. TwoSampleMR analysis

We will perform an MR analysis of body mass index (exposure - ieu-a-2) against coronary heart disease (outcome - ieu-a-7).

Obtain the data

```{r}
x <- clumpTophits(traits = "ieu-a-2")
sumset1 <- SummarySet(traits = "ieu-a-2", variants = x, tools = "mr")
sumset1
sumset2 <- SummarySet(traits="ieu-a-7", variants=x,tools ="mr")
sumset2

```

Set the Metadata and MR labels for each of your summary sets

```{r}
sumset1 <- setMetadata(sumset1, source = "IEUopenGWAS", traits = "ieu-a-2")
sumset1 <-setMRlabel(sumset1, mr_label = "exposure")

sumset2 <- setMetadata(sumset2,source = "IEUopenGWAS", traits = "ieu-a-7")
sumset2 <-setMRlabel(sumset2, mr_label = "outcome")

```


Check the Metadata and MR labels

```{r}
getMetadata(sumset1)
getMRlabel(sumset1)

getMetadata(sumset2)
getMRlabel(sumset2)
```


Create the DataSet object

```{r}
dataset <- DataSet(sumset1,sumset2) %>%
  overlapSNP(.) %>%
  harmoniseData(.,tolerance = 0.08,action = 2)
```

Convert dataset to TwoSampleMR format

```{r}
dataset_mr <- convertForTwoSampleMR(dataset)
```

Perform the MR analysis
```{r}
mr_result <-  merge(dataset_mr@summary_sets[[1]]@ss,dataset_mr@summary_sets[[2]]@ss, by = c("SNP", "mr_keep"))  %>%
  TwoSampleMR::mr(., method_list="mr_ivw")

mr_result

```


## 2. Regional genotype-phenotype map (WIP)

We need to first define a genomic range for the analysis to be performed. This is a known LD block on chromosome 22 (build hg19)

```{r}
genomicrange <- "22:43714200-44995307"

```

We need to scan OpenGWAS for traits that have a significant associations in this region

```{r}
id_list <- ieugwasr::phewas(pval = 5e-6, variants=genomicrange, batch="ukb-a")$id %>%
  unique()
```

Obtain data

```{r}
dataset <- gwasglue2::create_dataset(traits=id_list, variants=genomicrange) %>%
  gwasglue2::annotate(ld_matrix="1000genomes_EUR")
```

Next, for each trait we need to perform finemapping. This will involve using SusieR to identify credible sets in the region for the trait, and then generate the conditional summary datasets for each credible set. Explanation:

Credible set: If a trait is inferred to have e.g. 3 causal variants in the genomic region then there will be 3 credible sets. Each credible set is essentially a cluster of SNPs close by, it will include 1 or more SNPs that are candidates to be the causal variant.

Conditional summary datasets: Once credible sets are identified, we can try to estimate what the summary data results would have looked like for each credible set after accounting for all the other credible sets. So in this example there would be 3 conditional summary datasets, each with a single peak in a different location.

Use susieR to generate the credible sets for each dataset (e.g. some of them will have only 1, some will have more than 1. But all of them will have at least 1 because they have been selected to have an association in the region)

```{r}
dataset <- gwasglue2::annotate(dataset = ., credible_sets="susieR")
```

Finally, we can perform the colocalisation. Here we want to jointly colocalise across all the conditional summary datasets


```{r}
result <- dataset %>%
  hyprcoloc::hyprcoloc(dataset=.)
```
## Alternative coloc example (WIP)

```{r}
library(gwasglue2)
remotes::install_github('ritarasteiro/hyprcoloc', build_opts = c('--resave-data', '--no-manual'), build_vignettes = FALSE, dependencies = TRUE) 
remotes::install_github('ritarasteiro/susieR')
library(susieR)     # fork that takes Summaryset
library(hyprcoloc)  # fork that takes Dataset
```

We'll analyse the following IDs (chd - cardiac heart disease; ldl - low-density lipoprotein; hdl - high-density lipoprotein; trig - triglycerides and t2d - type 2 diabetes) around the HMGCR (3-hydroxy-3-methylglutaryl-CoA reductase) gene region. 


```{r}
chd_id <- "ieu-a-7"
ldl_id <- "ieu-b-110"
hdl_id <- "ieu-b-109"
trig_id <- "ieu-b-111"
t2d_id <- "ebi-a-GCST006867"
hmgcr_chrpos <- "5:74132993-75132993"
```


Create summarysets

```{r}
ids <- c(chd_id, ldl_id, hdl_id, trig_id, t2d_id)
# create s4 SummarySet objects (named sumset1...n) and fill metadata slot using the createSummarySets() in ieugwas_utils
for (i in seq_along(ids)){
  s <- createSummarySets(traits=ids[i],
                          variants = "5:74132993-75132993",
                          tools = c("finemap","coloc"),
                          source = "IEUopenGWAS",
                          ld_ref = "/project/data/reference/ld/EUR")                        
  assign(paste0("sumset",i) , s)
}
```

Create dataset and harmonise trait against trait, trait against LD matrix

```{r}
# create S4 DataSet object (hmgcr)
ds_hmgcr <- DataSet(sumset1,sumset2,sumset3,sumset4,sumset5) %>%
  overlapSNP(.) %>%
  harmoniseData(.,tolerance = 0.08,action = 1) %>%
  buildLDMatrix(., bfile = TRUE, plink_bin = "plink") %>%
  harmoniseLDMatrix(.) 
```

After summary sets are harmonised, marginalise each summary set independently and create a new dataset with all marginalised summary sets merged

```{r}
# do finemapping with susie
ntraits <- getLength(ds_hmgcr)
ds_hmgcr_marginalised <- lapply(1:ntraits, function(trait)
  {
    # Takes in 1 SS
    # Outputs 1 DS (with at least 1 SS)
    ds <- susie_rss(R = getLDMatrix(ds_hmgcr,trait), summaryset = getSummarySet(ds_hmgcr, trait))
}) 
susieR is using gwasglue2 SummarySet class object as input
WARNING: There is no credible sets for this trait (ieu-a-7), with the parameter values used. The summary statistics beta and se will not be marginalised.
susieR is using gwasglue2 SummarySet class object as input
susieR is using gwasglue2 SummarySet class object as input
susieR is using gwasglue2 SummarySet class object as input
WARNING: There is no credible sets for this trait (ieu-b-111), with the parameter values used. The summary statistics beta and se will not be marginalised.
susieR is using gwasglue2 SummarySet class object as input
WARNING: There is no credible sets for this trait (ebi-a-GCST006867), with the parameter values used. The summary statistics beta and se will not be marginalised.
```
Because not all datasets were marginalised, we are going to merge just the ones that were

```{r}
ds_hmgcr_marginalised <- merge_datasets(ds_hmgcr_marginalised[[2]], ds_hmgcr_marginalised[[3]])
```

Try hyprcoloc with raw datasets. (Use fork of hypercoloc that allows a dataset object to be provided)

```{r}
res_hmgcr <- hyprcoloc(dataset = ds_hmgcr)
hyprcoloc is using gwasglue2 DataSet class object as input
print(res_hmgcr)
$results
  iteration                      traits posterior_prob regional_prob
1         1          ieu-a-7, ieu-b-110         0.5940        0.6052
2         2 ieu-b-109, ebi-a-GCST006867         0.9934        0.9997
  candidate_snp posterior_explained_by_snp dropped_trait
1       rs12916                     0.9996            NA
2     rs2307111                     0.9932            NA
attr(,"class")
[1] "hyprcoloc"
```

Try hyprcoloc with marginalised datasets

```{r}
res_hmgcr_marginalised <- hyprcoloc(dataset = ds_hmgcr_marginalised)
hyprcoloc is using gwasglue2 DataSet class object as input
print(res_hmgcr_marginalised)
$results
  iteration                     traits posterior_prob regional_prob
1         1 ieu-b-110_L2, ieu-b-110_L3         0.7226        1.0000
2         2 ieu-b-110_L1, ieu-b-110_L4         0.3766        0.5999
  candidate_snp posterior_explained_by_snp dropped_trait
1    rs59992948                     0.0106            NA
2    rs16872674                     0.0087            NA
attr(,"class")
[1] "hyprcoloc"
```

Also try with these other regions:


```{r}
pcsk9_chrpos <- "1:55005221-56005221"
npc1l1_chrpos <- "7:44052134-45052134"
lpa_chrpos <- "6:155952514-165952514"
```


* PCSK9 (proprotein convertase subtilisin/kexin type 9) gene region

Create dataset and harmonise trait against trait, trait against LD matrix (note that we are not saving the SummarySet objects separately, but filling them directly to the DataSet).


```{r}
# create S4 DataSet object (pcsk9)
pcsk9_chrpos <- "1:55005221-56005221"
pcsk9 <- DataSet()
for (i in seq_along(ids)){
  s <- createSummarySets(traits=ids[i],
                          variants = "1:55005221-56005221",
                          tools = c("finemap","coloc"),
                          source = "IEUopenGWAS",
                          ld_ref = "/project/data/reference/ld/EUR")                        
 pcsk9@summary_sets[[i]] <- s
}
pcsk9 <-  pcsk9 %>%
  overlapSNP(.) %>%
  harmoniseData(.,tolerance = 0.08,action = 1) %>%
  buildLDMatrix(., bfile = TRUE, plink_bin = "plink") %>%
  harmoniseLDMatrix(.) 
```

Marginalise each summary set independently and create a new dataset with all marginalised summary sets merged

```{r}
# do finemapping with susie
ntraits <- getLength(pcsk9)
pcsk9_marginalised <- lapply(1:ntraits, function(trait)
  {
    # Takes in 1 SS
    # Outputs 1 DS (with at least 1 SS)
    ds <- susie_rss(R = getLDMatrix(pcsk9,trait), summaryset = getSummarySet(pcsk9, trait))
}) 
susieR is using gwasglue2 SummarySet class object as input
WARNING: There is no credible sets for this trait (ieu-a-7), with the parameter values used. The summary statistics beta and se will not be marginalised.
susieR is using gwasglue2 SummarySet class object as input
susieR is using gwasglue2 SummarySet class object as input
WARNING: There is no credible sets for this trait (ieu-b-109), with the parameter values used. The summary statistics beta and se will not be marginalised.
susieR is using gwasglue2 SummarySet class object as input
WARNING: There is no credible sets for this trait (ieu-b-111), with the parameter values used. The summary statistics beta and se will not be marginalised.
susieR is using gwasglue2 SummarySet class object as input
WARNING: There is no credible sets for this trait (ebi-a-GCST006867), with the parameter values used. The summary statistics beta and se will not be marginalised.
```
Because not all datasets were marginalised, we are going to merge just the ones that were

```{r}
ds_hmgcr_marginalised <- merge_datasets(ds_hmgcr_marginalised[[2]])
```


Try hyprcoloc with raw datasets. 

```{r}
res_pcsk9 <- hyprcoloc(dataset = pcsk9)
hyprcoloc is using gwasglue2 DataSet class object as input
print(res_pcsk9)
$results
  iteration traits posterior_prob regional_prob candidate_snp
1         1   None             NA        0.1472            NA
2         2   None             NA        0.0003            NA
3         3   None             NA        0.0001            NA
4         4   None             NA        0.0001            NA
  posterior_explained_by_snp    dropped_trait
1                         NA        ieu-b-110
2                         NA          ieu-a-7
3                         NA ebi-a-GCST006867
4                         NA        ieu-b-109
attr(,"class")
[1] "hyprcoloc"
```

Try hyprcoloc with marginalised datasets

```{r}
res_pcsk9_marginalised <- hyprcoloc(dataset =pcsk9_marginalised)
hyprcoloc is using gwasglue2 DataSet class object as input
print(res_pcsk9_marginalised)
$results
  iteration traits posterior_prob regional_prob candidate_snp
1         1   None             NA        0.6947            NA
2         2   None             NA        0.4759            NA
3         3   None             NA        0.3103            NA
4         4   None             NA        0.0017            NA
5         5   None             NA        0.0010            NA
6         6   None             NA        0.0001            NA
7         7   None             NA        0.0003            NA
8         8   None             NA        0.0001            NA
9         9   None             NA        0.0001            NA
  posterior_explained_by_snp dropped_trait
1                         NA  ieu-b-110_L6
2                         NA  ieu-b-110_L1
3                         NA  ieu-b-110_L5
4                         NA  ieu-b-110_L2
5                         NA  ieu-b-110_L7
6                         NA  ieu-b-110_L4
7                         NA  ieu-b-110_L8
8                         NA  ieu-b-110_L3
9                         NA  ieu-b-110_L9
attr(,"class")
[1] "hyprcoloc"
```

* NPC1L1 (NPC1 like intracellular cholesterol transporter 1) gene region
Create dataset and harmonise trait against trait, trait against LD matrix

```{r}
# create S4 DataSet object (npc1l1)
npc1l1_chrpos <- "7:44052134-45052134"
npc1l1 <- DataSet()
for (i in seq_along(ids)){
  s <- createSummarySets(traits=ids[i],
                          variants = "7:44052134-45052134",
                          tools = c("finemap","coloc"),
                          source = "IEUopenGWAS",
                          ld_ref = "/project/data/reference/ld/EUR")                        
 npc1l1@summary_sets[[i]] <- s
}
npc1l1 <-  npc1l1 %>%
  overlapSNP(.) %>%
  harmoniseData(.,tolerance = 0.08,action = 1) %>%
  buildLDMatrix(., bfile = TRUE, plink_bin = "plink") %>%
  harmoniseLDMatrix(.) 
```

Marginalise each summary set independently and create a new dataset with all marginalised summary sets merged

```{r}
ntraits <- getLength( npc1l1)
npc1l1_marginalised <- lapply(1:ntraits, function(trait)
  {
    # Takes in 1 SS
    # Outputs 1 DS (with at least 1 SS)
    ds <- susie_rss(R = getLDMatrix(npc1l1,trait), summaryset = getSummarySet( npc1l1, trait))
}) 
susieR is using gwasglue2 SummarySet class object as input
WARNING: There is no credible sets for this trait (ieu-a-7), with the parameter values used. The summary statistics beta and se will not be marginalised.
susieR is using gwasglue2 SummarySet class object as input
susieR is using gwasglue2 SummarySet class object as input
susieR is using gwasglue2 SummarySet class object as input
susieR is using gwasglue2 SummarySet class object as input
```
Because not all datasets were marginalised, we are going to merge just the ones that were

```{r}
npc1l1_marginalised <- merge_datasets(npc1l1_marginalised[[2]],npc1l1_marginalised[[3]],npc1l1_marginalised[[4]],npc1l1_marginalised[[5]])
```


Try hyprcoloc with raw datasets. 

```{r}
res_npc1l1 <- hyprcoloc(dataset = npc1l1)
hyprcoloc is using gwasglue2 DataSet class object as input
print(res_npc1l1)
$results
  iteration traits posterior_prob regional_prob candidate_snp
1         1   None             NA        0.0348            NA
2         2   None             NA        0.2827            NA
3         3   None             NA        0.0370            NA
4         4   None             NA        0.0041            NA
  posterior_explained_by_snp    dropped_trait
1                         NA        ieu-b-110
2                         NA        ieu-b-111
3                         NA ebi-a-GCST006867
4                         NA        ieu-b-109
attr(,"class")
[1] "hyprcoloc
```
Try hyprcoloc with marginalised datasets
```{r}
res_npc1l1_marginalised <- hyprcoloc(dataset =npc1l1_marginalised)
hyprcoloc is using gwasglue2 DataSet class object as input
print(res_npc1l1_marginalised)
$results
  iteration traits posterior_prob regional_prob candidate_snp
1         1   None             NA        0.0402            NA
2         2   None             NA        0.0577            NA
3         3   None             NA        0.0082            NA
4         4   None             NA        0.0027            NA
5         5   None             NA        0.0012            NA
6         6   None             NA        0.0009            NA
  posterior_explained_by_snp       dropped_trait
1                         NA        ieu-b-111_L1
2                         NA ebi-a-GCST006867_L1
3                         NA        ieu-b-110_L2
4                         NA        ieu-b-109_L1
5                         NA        ieu-b-111_L2
6                         NA        ieu-b-110_L1
attr(,"class")
[1] "hyprcoloc"
```
* LPA (Lipoprotein(A)) gene region
Create dataset and harmonise trait against trait, trait against LD matrix
```{r}
# create S4 DataSet object (lpa)
lpa_chrpos <- "6:155952514-165952514"
lpa <- DataSet()
for (i in seq_along(ids)){
  s <- createSummarySets(traits=ids[i],
                          variants = "6:155952514-165952514",
                          tools = c("finemap","coloc"),
                          source = "IEUopenGWAS",
                          ld_ref = "/project/data/reference/ld/EUR")                        
 lpa@summary_sets[[i]] <- s
}
lpa <-  lpa %>%
  overlapSNP(.) %>%
  harmoniseData(.,tolerance = 0.08,action = 1) %>%
  buildLDMatrix(., bfile = TRUE, plink_bin = "plink") %>%
  harmoniseLDMatrix(.) 
```
Marginalise each summary set independently and create a new dataset with all marginalised summary sets merged
```{r}
ntraits <- getLength(lpa)
npc1l1_marginalised <- lapply(1:ntraits, function(trait)
  {
    # Takes in 1 SS
    # Outputs 1 DS (with at least 1 SS)
    ds <- susie_rss(R = getLDMatrix(lpa,trait), summaryset = getSummarySet(lpa, trait))
}) 
lpa_marginalised <- merge_datasets()
```
Because not all datasets were marginalised, we are going to merge just the ones that were
```{r}
```
Try hyprcoloc with raw datasets. 
```{r}
res_lpa <- hyprcoloc(dataset = nlpa)
print(res_lpa)
```
Try hyprcoloc with marginalised datasets
```{r}
res_lpa_marginalised <- hyprcoloc(dataset =lpa_marginalised)
print(res_lpa_marginalised)
```
