---
title: "Tutorial"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```


```{r  eval= FALSE}
remotes::install_github('ritarasteiro/hyprcoloc', build_opts = c('--resave-data', '--no-manual'), build_vignettes = TRUE) 
remotes::install_github('ritarasteiro/susieR')

```

```{r setup}
library(gwasglue2)
library(TwoSampleMR)
library(ieugwasr)
library(susieR)     # fork that takes Summaryset
library(hyprcoloc)  # fork that takes Dataset
devtools::load_all("../") # this was added just for development
```

There are two usecases for this tutorial 

1. Performing a basic TwoSampleMR analysis
2. Generating a regional genotype-phenotype map (Million Veteran's Programme analysis example)


#### This tutorial is still work in progress. It reflects an earlier draft of gwasglue2.

## 1. TwoSampleMR analysis

We will perform an MR analysis of body mass index (exposure - ieu-a-2) against coronary heart disease (outcome - ieu-a-7).

Obtain the data

```{r}
x <- ieugwasr::tophits("ieu-a-2")$rsid
d1 <- ieugwasr::associations(variants = x, id = "ieu-a-2")
d2 <- ieugwasr::associations(variants = x, id = "ieu-a-7")
```

Obtain the metadata and create a metadata object

```{r}
m1 <- ieugwasr::gwasinfo( "ieu-a-2")
m2 <- ieugwasr::gwasinfo( "ieu-a-7")
meta1 <-create_metadata(m1)
meta2 <-create_metadata(m2)

```
Construct the Summary Sets and set the tool and metadata (it uses ieugwasr::gwasinfo() with the source and id arguments)

```{r}
sumset1 <- create_summaryset(d1, metadata=meta1, tools = "mr")
sumset2 <- create_summaryset(d2, metadata=meta2, tools ="mr")

```

Set the MR labels for each of your summary sets

```{r}
sumset1 <-setMRlabel(sumset1, mr_label = "exposure")
sumset2 <-setMRlabel(sumset2, mr_label = "outcome")

```


Check the Metadata and MR labels

```{r}
getMetadata(sumset1)
getMRlabel(sumset1)

getMetadata(sumset2)
getMRlabel(sumset2)
```


Create the DataSet object

```{r}
dataset <- DataSet(sumset1,sumset2) %>%
  overlapVariants(., strand = "forward") %>%
  harmoniseData(.,tolerance = 0.08, action = 1)
```

Convert dataset to TwoSampleMR format

```{r}
dataset_mr <- convertForTwoSampleMR(dataset)
```

Perform the MR analysis
```{r}
mr_result <-  merge(dataset_mr@summary_sets[[1]]@ss,dataset_mr@summary_sets[[2]]@ss, by = c("SNP", "mr_keep"))  %>%
  TwoSampleMR::mr(., method_list="mr_ivw")

mr_result

```


## 2. Regional genotype-phenotype map (WIP)

We need to first define a genomic range for the analysis to be performed. This is a known LD block on chromosome 22 (build hg19)

```{r}
genomicrange <- "22:43714200-44995307"

```

We need to scan OpenGWAS for traits that have a significant associations in this region

```{r}
id_list <- ieugwasr::phewas(pval = 5e-6, variants=genomicrange, batch="ukb-a")$id %>%
  unique()
```

First we will set the path to where the reference population plink files are located to build the LD matrix.

```{r}
#ld_ref <- "/project/data/reference/ld/EUR"
ld_ref <- "data/ld/EUR"
```

and then obtain the data

```{r}
# Get GWAS summary statistics
data <- lapply(seq_along(id_list), function(i){
    s <- ieugwasr::associations(variants = genomicrange, id = id_list[i])
})

# get metadata and create metadata objects
metadata <- lapply(seq_along(id_list), function(i){
  m <- create_metadata(ieugwasr::gwasinfo(id_list[i])) 
})


# create gwasglue2 DataSet object 
  dataset <- create_dataset(data, metadata = metadata, tools = c("finemap","coloc"), harmonise = TRUE, tolerance = 0.08, action = 1) %>% 
  harmonise_ld(., ld_ref = ld_ref, bfile = TRUE, plink_bin = "plink")
```

Next, for each trait we need to perform finemapping. This will involve using SusieR to identify credible sets in the region for the trait, and then generate the conditional summary datasets for each credible set. Explanation:

Credible set: If a trait is inferred to have e.g. 3 causal variants in the genomic region then there will be 3 credible sets. Each credible set is essentially a cluster of SNPs close by, it will include 1 or more SNPs that are candidates to be the causal variant.

Conditional summary datasets: Once credible sets are identified, we can try to estimate what the summary data results would have looked like for each credible set after accounting for all the other credible sets. So in this example there would be 3 conditional summary datasets, each with a single peak in a different location.

Use susieR to generate the credible sets for each dataset (e.g. some of them will have only 1, some will have more than 1. But all of them will have at least 1 because they have been selected to have an association in the region)

TODO: Explain how we marginalise the summarysets using susieR Approximate Bayes Factors.

After summary sets are harmonised, marginalise each summary set independently and create a new dataset with all marginalised summary sets merged

```{r}
# do finemapping with susie
ntraits <- getLength(dataset)
dataset_marginalised <- lapply(1:ntraits, function(trait)
  {
    # Takes in 1 SS
    # Outputs 1 DS (with at least 1 SS)
    ds <- susie_rss(R = getLDMatrix(dataset), summaryset = getSummarySet(dataset, trait))
})
```

 Now we are going to merge the datasets. Note that the `merge_datasets()` function is just going to merge the marginalised datasets. 

```{r}
dataset_marginalised <- merge_datasets(dataset_marginalised)

``` 

 Try hyprcoloc with raw datasets. (Use fork of hypercoloc that allows a dataset object to be provided)

```{r} 

res_dataset <- hyprcoloc(dataset = dataset) 
#hyprcoloc is using gwasglue2 DataSet class object as input
print(res_dataset)

```

Try hyprcoloc with marginalised datasets

```{r}

res_dataset_marginalised <- hyprcoloc(dataset = dataset_marginalised)
print(res_dataset_marginalised)
```

**NOTE:** In this example there was no colocalisation. 
