---
title: "Tutorial"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

```{r setup}
library(gwasglue2)
library(TwoSampleMR)
library(ieugwasr)
library(hyprcoloc)
library(susieR)
```

There are two usecases for this tutorial 

1. Performing a basic TwoSampleMR analysis
2. Generating a regional genotype-phenotype map (Million Veteran's Programme analysis example)


#### This tutorial is still work in progress. It reflects an earlier draft of gwasglue2.

## 1. TwoSampleMR analysis

We will perform an MR analysis of body mass index (exposure - ieu-a-2) against coronary heart disease (outcome - ieu-a-7).

Obtain the data

```{r}
x <- clumpTophits(traits = "ieu-a-2")
sumset1 <- SummarySet(traits = "ieu-a-2", variants = x, tools = "mr")
sumset1
sumset2 <- SummarySet(traits="ieu-a-7", variants=x,tools ="mr")
sumset2

```

Set the Metadata and MR labels for each of your summary sets

```{r}
sumset1 <- setMetadata(sumset1, source = "IEUopenGWAS", traits = "ieu-a-2")
sumset1 <-setMRlabel(sumset1, mr_label = "exposure")

sumset2 <- setMetadata(sumset2,source = "IEUopenGWAS", traits = "ieu-a-7")
sumset2 <-setMRlabel(sumset2, mr_label = "outcome")

```


Check the Metadata and MR labels

```{r}
getMetadata(sumset1)
getMRlabel(sumset1)

getMetadata(sumset2)
getMRlabel(sumset2)
```


Create the DataSet object

```{r}
dataset <- DataSet(sumset1,sumset2) %>%
  overlapSNP(.) %>%
  harmoniseData(.,tolerance = 0.08,action = 2)
```

Convert dataset to TwoSampleMR format

```{r}
dataset_mr <- convertForTwoSampleMR(dataset)
```

Perform the MR analysis
```{r}
mr_result <-  merge(dataset_mr@summary_sets[[1]]@ss,dataset_mr@summary_sets[[2]]@ss, by = c("SNP", "mr_keep"))  %>%
  TwoSampleMR::mr(., method_list="mr_ivw")

mr_result

```


## 2. Regional genotype-phenotype map (WIP)

We need to first define a genomic range for the analysis to be performed. This is a known LD block on chromosome 22 (build hg19)

```{r}
genomicrange <- "22:43714200-44995307"

```

We need to scan OpenGWAS for traits that have a significant associations in this region

```{r}
id_list <- ieugwasr::phewas(pval = 5e-6, variants=genomicrange, batch="ukb-a")$id %>%
  unique()
```

Obtain data

```{r}
dataset <- gwasglue2::create_dataset(traits=id_list, variants=genomicrange) %>%
  gwasglue2::annotate(ld_matrix="1000genomes_EUR")
```

Next, for each trait we need to perform finemapping. This will involve using SusieR to identify credible sets in the region for the trait, and then generate the conditional summary datasets for each credible set. Explanation:

Credible set: If a trait is inferred to have e.g. 3 causal variants in the genomic region then there will be 3 credible sets. Each credible set is essentially a cluster of SNPs close by, it will include 1 or more SNPs that are candidates to be the causal variant.

Conditional summary datasets: Once credible sets are identified, we can try to estimate what the summary data results would have looked like for each credible set after accounting for all the other credible sets. So in this example there would be 3 conditional summary datasets, each with a single peak in a different location.

Use susieR to generate the credible sets for each dataset (e.g. some of them will have only 1, some will have more than 1. But all of them will have at least 1 because they have been selected to have an association in the region)

```{r}
dataset <- gwasglue2::annotate(dataset = ., credible_sets="susieR")
```

Finally, we can perform the colocalisation. Here we want to jointly colocalise across all the conditional summary datasets


```{r}
result <- dataset %>%
  hyprcoloc::hyprcoloc(dataset=.)
```

## Alternative coloc example

```{r}
library(gwasglue2)
library(susieR)     # fork that takes Summaryset
library(hyprcoloc)  # fork that takes Dataset
```

We'll analyse the following IDs around the HMGCR gene region

```{r}
chd_id <- "ieu-a-7"
ldl_id <- "ieu-b-110"
hdl_id <- "ieu-b-109"
trig_id <- "ieu-b-111"
t2d_id <- "ebi-a-GCST006867"
hmgcr_chrpos <- "5:74132993-75132993"
```

Create dataset

```{r}
dataset <- gwasglue2::create_dataset(traits=c(chd_id, ldl_id, hdl_id, trig_id, t2d_id), variants=hmgcr_chrpos) %>%
  gwasglue2::annotate(ld_matrix="1000genomes_EUR")
```

After summary sets are harmonised, marginalise each summary set independently and create a new dataset with all marginalised summary sets

```{r}
ds_marginalised <- lapply(getTraits(ds), function(trait){
    # Takes in 1 SS
    # Outputs 1 DS (with at least 1 SS)
    susie_rss(getData(ds, trait))
}) %>%
    gwasglue2::merge_datasets(.)
```

Try hyprcoloc with raw datasets. (Use fork of hypercoloc that allows a dataset object to be provided)

```{r}
res_simple <- hyprcoloc(ds)
```

Try hyprcoloc with marginalised datasets

```{r}
res_marginalised <- hypercoloc(ds_marginalised)
```

Also try with these other regions:

```{r}
pcsk9_chrpos <- "1:55005221-56005221"
npc1l1_chrpos <- "7:44052134-45052134"
lpa_chrpos <- "6:155952514-165952514"
```
