<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="gwasglue2">
<title>Processing: liftover • gwasglue2</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Processing: liftover">
<meta property="og:description" content="gwasglue2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">gwasglue2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/Strategy.html">Strategy</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-howtos">HOWTOs</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-howtos">
    <a class="dropdown-item" href="../articles/input.html">Input</a>
    <a class="dropdown-item" href="../articles/processing.html">Processing</a>
    <a class="dropdown-item" href="../articles/output.html">Output</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/mr.html">MR analysis</a>
    <a class="dropdown-item" href="../articles/finemap_coloc1.html">Regional genotype-phenotype map</a>
    <a class="dropdown-item" href="../articles/ldsc.html">LD score regression</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-yaml">YAML</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-yaml">
    <a class="dropdown-item" href="../articles/yaml_tutorials.html">YAML tutorials</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MRCIEU/gwasglue2/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="processing-liftover_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Processing: liftover</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/gwasglue2/blob/HEAD/vignettes/processing-liftover.Rmd" class="external-link"><code>vignettes/processing-liftover.Rmd</code></a></small>
      <div class="d-none name"><code>processing-liftover.Rmd</code></div>
    </div>

    
    
<p><strong>gwasglue2</strong> has a <code><a href="../reference/liftover.html">liftover()</a></code> function to remap genomic coordinates from different genome assemblies to the same reference build. This function works on the <code>SummarySet</code> object and uses a chain file to map the coordinates. The chain file is a text file that contains the mapping between two genome assemblies. The chain file can be provided by the user or downloaded directly by the <strong>gwasglue2</strong> <code><a href="../reference/download_chainfile.html">download_chainfile()</a></code> or <code><a href="../reference/liftover.html">liftover()</a></code> functions.</p>
<p><strong>gwasglue2</strong> uses the chain files and Human genome nomenclature of <a href="https://www.ensembl.org/Homo_sapiens/" class="external-link">ENSEMBL</a> (“NCBI34”, “NCBI35”, “NCBI36”, “GRCh37” and “GRCh38”).</p>
<p><br><br> ## Setup In addition to <strong>gwasglue2</strong>, we’ll load the following package:</p>
<ul>
<li>
<a href="https://mrcieu.github.io/ieugwasr/" class="external-link"><strong>ieugwasr</strong></a> - R package for accessing IEU GWAS database</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/gwasglue2" class="external-link">gwasglue2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/ieugwasr" class="external-link">ieugwasr</a></span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/load_all.html" class="external-link">load_all</a></span><span class="op">(</span><span class="st">"../"</span><span class="op">)</span> <span class="co"># this was added just for development</span></span></code></pre></div>
<p><br><br> ## Functions * <code><a href="../reference/create_summaryset.html">create_summaryset()</a></code>, using the <code>build</code> argument. * <code><a href="../reference/download_chainfile.html">download_chainfile()</a></code>: downloads Human genomes chain files from ENSEMBL. * <code><a href="../reference/liftover.html">liftover()</a></code>: remaps the genomic coordinates from the GWAS summary data on the <code>SummarySet</code> to another genome assembly using a chain file.</p>
<p><br><br></p>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>We will use GWAS summary data obtained from IEU OpenGWAS through the <strong>ieugwasr</strong> package (see <a href="input-opengwas.html">here</a> for details). The data was built under the “GRCh37” genome assembly and we want to remap it to “GRCh378”.</p>
<p>First, we create a SummarySet object that we name <code>sumset</code> providing in the <code>build</code> argument the genome assembly used to build the GWAS summary data. We can check it in the metadata.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/associations.html" class="external-link">associations</a></span><span class="op">(</span>variants <span class="op">=</span> <span class="st">"5:74132993-75132993"</span>, id <span class="op">=</span> <span class="st">"ukb-d-I9_IHD"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sumset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_summaryset.html">create_summaryset</a></span><span class="op">(</span><span class="va">data</span>, build <span class="op">=</span> <span class="st">"GRCh37"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/getMetadata-methods.html">getMetadata</a></span><span class="op">(</span><span class="va">sumset</span><span class="op">)</span></span></code></pre></div>
<p>Then we perform the liftover. We can first download the chain file followed by the <code><a href="../reference/liftover.html">liftover()</a></code>. In this case we need to give the genome assemblies to “lift from” and to “lift to” :</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumset_lift</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_chainfile.html">download_chainfile</a></span><span class="op">(</span>from <span class="op">=</span> <span class="st">"GRCh37"</span>, to <span class="op">=</span> <span class="st">"GRCh38"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                <span class="fu"><a href="../reference/liftover.html">liftover</a></span><span class="op">(</span><span class="va">sumset</span>, chainfile <span class="op">=</span> <span class="va">.</span>, to <span class="op">=</span> <span class="st">"GRCh38"</span><span class="op">)</span></span></code></pre></div>
<p>Or simply provide directly to <code><a href="../reference/liftover.html">liftover()</a></code> the genome assembly to “lift to”. In this case, the function will check the <code>build</code> to “lift from” in the metadata of <code>sumset</code> and download the chain file.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumset1_lift</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/liftover.html">liftover</a></span><span class="op">(</span><span class="va">sumset</span>, to <span class="op">=</span> <span class="st">"GRCh38"</span><span class="op">)</span></span></code></pre></div>
</div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Rita Rasteiro, Gibran Hemani, Christopher Woods.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
