<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="gwasglue2">
<title>Tutorial • gwasglue2</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial">
<meta property="og:description" content="gwasglue2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">gwasglue2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Strategy.html">Strategy</a>
    <a class="dropdown-item" href="../articles/Tutorial.html">Tutorial</a>
    <a class="dropdown-item" href="../articles/Tutorial2.html">Tutorial2</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MRCIEU/gwasglue2/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="Tutorial_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Tutorial</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/gwasglue2/blob/HEAD/vignettes/Tutorial.Rmd" class="external-link"><code>vignettes/Tutorial.Rmd</code></a></small>
      <div class="d-none name"><code>Tutorial.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'ritarasteiro/hyprcoloc'</span>, build_opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'--resave-data'</span>, <span class="st">'--no-manual'</span><span class="op">)</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'ritarasteiro/susieR'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/gwasglue2" class="external-link">gwasglue2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TwoSampleMR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/ieugwasr" class="external-link">ieugwasr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/susieR" class="external-link">susieR</a></span><span class="op">)</span>     <span class="co"># fork that takes Summaryset</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">hyprcoloc</span><span class="op">)</span>  <span class="co"># fork that takes Dataset</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/load_all.html" class="external-link">load_all</a></span><span class="op">(</span><span class="st">"../"</span><span class="op">)</span> <span class="co"># this was added just for development</span></span></code></pre></div>
<p>There are two usecases for this tutorial</p>
<ol style="list-style-type: decimal">
<li>Performing a basic TwoSampleMR analysis</li>
<li>Generating a regional genotype-phenotype map (Million Veteran’s Programme analysis example)</li>
</ol>
<div class="section level4">
<h4 id="this-tutorial-is-still-work-in-progress--it-reflects-an-earlier-draft-of-gwasglue2-">This tutorial is still work in progress. It reflects an earlier draft of gwasglue2.<a class="anchor" aria-label="anchor" href="#this-tutorial-is-still-work-in-progress--it-reflects-an-earlier-draft-of-gwasglue2-"></a>
</h4>
</div>
<div class="section level2">
<h2 id="twosamplemr-analysis">1. TwoSampleMR analysis<a class="anchor" aria-label="anchor" href="#twosamplemr-analysis"></a>
</h2>
<p>We will perform an MR analysis of body mass index (exposure - ieu-a-2) against coronary heart disease (outcome - ieu-a-7).</p>
<p>Obtain the data</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/tophits.html" class="external-link">tophits</a></span><span class="op">(</span><span class="st">"ieu-a-2"</span><span class="op">)</span><span class="op">$</span><span class="va">rsid</span></span>
<span><span class="va">d1</span> <span class="op">&lt;-</span> <span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/associations.html" class="external-link">associations</a></span><span class="op">(</span>variants <span class="op">=</span> <span class="va">x</span>, id <span class="op">=</span> <span class="st">"ieu-a-2"</span><span class="op">)</span></span>
<span><span class="va">d2</span> <span class="op">&lt;-</span> <span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/associations.html" class="external-link">associations</a></span><span class="op">(</span>variants <span class="op">=</span> <span class="va">x</span>, id <span class="op">=</span> <span class="st">"ieu-a-7"</span><span class="op">)</span></span></code></pre></div>
<p>Obtain the metadata and create a metadata object</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m1</span> <span class="op">&lt;-</span> <span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/gwasinfo.html" class="external-link">gwasinfo</a></span><span class="op">(</span> <span class="st">"ieu-a-2"</span><span class="op">)</span></span>
<span><span class="va">m2</span> <span class="op">&lt;-</span> <span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/gwasinfo.html" class="external-link">gwasinfo</a></span><span class="op">(</span> <span class="st">"ieu-a-7"</span><span class="op">)</span></span>
<span><span class="va">meta1</span> <span class="op">&lt;-</span><span class="fu"><a href="../reference/create_metadata.html">create_metadata</a></span><span class="op">(</span><span class="va">m1</span><span class="op">)</span></span>
<span><span class="va">meta2</span> <span class="op">&lt;-</span><span class="fu"><a href="../reference/create_metadata.html">create_metadata</a></span><span class="op">(</span><span class="va">m2</span><span class="op">)</span></span></code></pre></div>
<p>Construct the Summary Sets and set the tool and metadata (it uses ieugwasr::gwasinfo() with the source and id arguments)</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumset1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_summaryset.html">create_summaryset</a></span><span class="op">(</span><span class="va">d1</span>, metadata<span class="op">=</span><span class="va">meta1</span>, tools <span class="op">=</span> <span class="st">"mr"</span><span class="op">)</span></span>
<span><span class="va">sumset2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_summaryset.html">create_summaryset</a></span><span class="op">(</span><span class="va">d2</span>, metadata<span class="op">=</span><span class="va">meta2</span>, tools <span class="op">=</span><span class="st">"mr"</span><span class="op">)</span></span></code></pre></div>
<p>Set the MR labels for each of your summary sets</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumset1</span> <span class="op">&lt;-</span><span class="fu"><a href="../reference/setMRlabel-methods.html">setMRlabel</a></span><span class="op">(</span><span class="va">sumset1</span>, mr_label <span class="op">=</span> <span class="st">"exposure"</span><span class="op">)</span></span>
<span><span class="va">sumset2</span> <span class="op">&lt;-</span><span class="fu"><a href="../reference/setMRlabel-methods.html">setMRlabel</a></span><span class="op">(</span><span class="va">sumset2</span>, mr_label <span class="op">=</span> <span class="st">"outcome"</span><span class="op">)</span></span></code></pre></div>
<p>Check the Metadata and MR labels</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getMetadata-methods.html">getMetadata</a></span><span class="op">(</span><span class="va">sumset1</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/getMRlabel-methods.html">getMRlabel</a></span><span class="op">(</span><span class="va">sumset1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/getMetadata-methods.html">getMetadata</a></span><span class="op">(</span><span class="va">sumset2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/getMRlabel-methods.html">getMRlabel</a></span><span class="op">(</span><span class="va">sumset2</span><span class="op">)</span></span></code></pre></div>
<p>Create the DataSet object</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DataSet.html">DataSet</a></span><span class="op">(</span><span class="va">sumset1</span>,<span class="va">sumset2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/overlapVariants-methods.html">overlapVariants</a></span><span class="op">(</span><span class="va">.</span>, strand <span class="op">=</span> <span class="st">"forward"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/harmoniseData-methods.html">harmoniseData</a></span><span class="op">(</span><span class="va">.</span>,tolerance <span class="op">=</span> <span class="fl">0.08</span>, action <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>Convert dataset to TwoSampleMR format</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset_mr</span> <span class="op">&lt;-</span> <span class="fu">convertForTwoSampleMR</span><span class="op">(</span><span class="va">dataset</span><span class="op">)</span></span></code></pre></div>
<p>Perform the MR analysis</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mr_result</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">dataset_mr</span><span class="op">@</span><span class="va">summary_sets</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">ss</span>,<span class="va">dataset_mr</span><span class="op">@</span><span class="va">summary_sets</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">ss</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNP"</span>, <span class="st">"mr_keep"</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">TwoSampleMR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/mr.html" class="external-link">mr</a></span><span class="op">(</span><span class="va">.</span>, method_list<span class="op">=</span><span class="st">"mr_ivw"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mr_result</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="regional-genotype-phenotype-map-wip">2. Regional genotype-phenotype map (WIP)<a class="anchor" aria-label="anchor" href="#regional-genotype-phenotype-map-wip"></a>
</h2>
<p>We need to first define a genomic range for the analysis to be performed. This is a known LD block on chromosome 22 (build hg19)</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomicrange</span> <span class="op">&lt;-</span> <span class="st">"22:43714200-44995307"</span></span></code></pre></div>
<p>We need to scan OpenGWAS for traits that have a significant associations in this region</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">id_list</span> <span class="op">&lt;-</span> <span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/phewas.html" class="external-link">phewas</a></span><span class="op">(</span>pval <span class="op">=</span> <span class="fl">5e-6</span>, variants<span class="op">=</span><span class="va">genomicrange</span>, batch<span class="op">=</span><span class="st">"ukb-a"</span><span class="op">)</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>First we will set the path to where the reference population plink files are located to build the LD matrix.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#ld_ref &lt;- "/project/data/reference/ld/EUR"</span></span>
<span><span class="va">ld_ref</span> <span class="op">&lt;-</span> <span class="st">"data/ld/EUR"</span></span></code></pre></div>
<p>and then obtain the data</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get GWAS summary statistics</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">id_list</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">s</span> <span class="op">&lt;-</span> <span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/associations.html" class="external-link">associations</a></span><span class="op">(</span>variants <span class="op">=</span> <span class="va">genomicrange</span>, id <span class="op">=</span> <span class="va">id_list</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get metadata and create metadata objects</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">id_list</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_metadata.html">create_metadata</a></span><span class="op">(</span><span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/gwasinfo.html" class="external-link">gwasinfo</a></span><span class="op">(</span><span class="va">id_list</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># create gwasglue2 DataSet object </span></span>
<span>  <span class="va">dataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_dataset.html">create_dataset</a></span><span class="op">(</span><span class="va">data</span>, metadata <span class="op">=</span> <span class="va">metadata</span>, tools <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"finemap"</span>,<span class="st">"coloc"</span><span class="op">)</span>, harmonise <span class="op">=</span> <span class="cn">TRUE</span>, tolerance <span class="op">=</span> <span class="fl">0.08</span>, action <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/harmonise_ld.html">harmonise_ld</a></span><span class="op">(</span><span class="va">.</span>, ld_ref <span class="op">=</span> <span class="va">ld_ref</span>, bfile <span class="op">=</span> <span class="cn">TRUE</span>, plink_bin <span class="op">=</span> <span class="st">"plink"</span><span class="op">)</span></span></code></pre></div>
<p>Next, for each trait we need to perform finemapping. This will involve using SusieR to identify credible sets in the region for the trait, and then generate the conditional summary datasets for each credible set. Explanation:</p>
<p>Credible set: If a trait is inferred to have e.g. 3 causal variants in the genomic region then there will be 3 credible sets. Each credible set is essentially a cluster of SNPs close by, it will include 1 or more SNPs that are candidates to be the causal variant.</p>
<p>Conditional summary datasets: Once credible sets are identified, we can try to estimate what the summary data results would have looked like for each credible set after accounting for all the other credible sets. So in this example there would be 3 conditional summary datasets, each with a single peak in a different location.</p>
<p>Use susieR to generate the credible sets for each dataset (e.g. some of them will have only 1, some will have more than 1. But all of them will have at least 1 because they have been selected to have an association in the region)</p>
<p>TODO: Explain how we marginalise the summarysets using susieR Approximate Bayes Factors.</p>
<p>After summary sets are harmonised, marginalise each summary set independently and create a new dataset with all marginalised summary sets merged</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># do finemapping with susie</span></span>
<span><span class="va">ntraits</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getLength-methods.html">getLength</a></span><span class="op">(</span><span class="va">dataset</span><span class="op">)</span></span>
<span><span class="va">dataset_marginalised</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">ntraits</span>, <span class="kw">function</span><span class="op">(</span><span class="va">trait</span><span class="op">)</span></span>
<span>  <span class="op">{</span></span>
<span>    <span class="co"># Takes in 1 SS</span></span>
<span>    <span class="co"># Outputs 1 DS (with at least 1 SS)</span></span>
<span>    <span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/susieR/man/susie_rss.html" class="external-link">susie_rss</a></span><span class="op">(</span>R <span class="op">=</span> <span class="fu">getLDMatrix</span><span class="op">(</span><span class="va">dataset</span><span class="op">)</span>, summaryset <span class="op">=</span> <span class="fu"><a href="../reference/getSummarySet-methods.html">getSummarySet</a></span><span class="op">(</span><span class="va">dataset</span>, <span class="va">trait</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Now we are going to merge the datasets. Note that the <code><a href="../reference/merge_datasets.html">merge_datasets()</a></code> function is just going to merge the marginalised datasets.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset_marginalised</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_datasets.html">merge_datasets</a></span><span class="op">(</span><span class="va">dataset_marginalised</span><span class="op">)</span></span></code></pre></div>
<p>Try hyprcoloc with raw datasets. (Use fork of hypercoloc that allows a dataset object to be provided)</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">res_dataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/hyprcoloc/man/hyprcoloc.html" class="external-link">hyprcoloc</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="va">dataset</span><span class="op">)</span> </span>
<span><span class="co">#hyprcoloc is using gwasglue2 DataSet class object as input</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res_dataset</span><span class="op">)</span></span></code></pre></div>
<p>Try hyprcoloc with marginalised datasets</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">res_dataset_marginalised</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/hyprcoloc/man/hyprcoloc.html" class="external-link">hyprcoloc</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="va">dataset_marginalised</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res_dataset_marginalised</span><span class="op">)</span></span></code></pre></div>
<p><strong>NOTE:</strong> In this example there was no colocalisation.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Gibran Hemani, Rita Rasteiro, Christopher Woods.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
