<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="gwasglue2">
<title>Tutorial2 • gwasglue2</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial2">
<meta property="og:description" content="gwasglue2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">gwasglue2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Strategy.html">Strategy</a>
    <a class="dropdown-item" href="../articles/Tutorial.html">Tutorial</a>
    <a class="dropdown-item" href="../articles/Tutorial2.html">Tutorial2</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MRCIEU/gwasglue2/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="Tutorial2_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Tutorial2</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/gwasglue2/blob/HEAD/vignettes/Tutorial2.Rmd" class="external-link"><code>vignettes/Tutorial2.Rmd</code></a></small>
      <div class="d-none name"><code>Tutorial2.Rmd</code></div>
    </div>

    
    
<div class="section level4">
<h4 id="this-tutorial-is-still-work-in-progress--it-reflects-an-earlier-draft-of-gwasglue2-">This tutorial is still work in progress. It reflects an earlier draft of gwasglue2.<a class="anchor" aria-label="anchor" href="#this-tutorial-is-still-work-in-progress--it-reflects-an-earlier-draft-of-gwasglue2-"></a>
</h4>
</div>
<div class="section level2">
<h2 id="alternative-regional-genotype-phenotype-map-wip">1. Alternative regional genotype-phenotype map (WIP)<a class="anchor" aria-label="anchor" href="#alternative-regional-genotype-phenotype-map-wip"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'ritarasteiro/hyprcoloc'</span>, build_opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'--resave-data'</span>, <span class="st">'--no-manual'</span><span class="op">)</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'ritarasteiro/susieR'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/gwasglue2" class="external-link">gwasglue2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/susieR" class="external-link">susieR</a></span><span class="op">)</span>     <span class="co"># fork that takes Summaryset</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">hyprcoloc</span><span class="op">)</span>  <span class="co"># fork that takes Dataset</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span></code></pre></div>
<p>We’ll analyse the following IDs (chd - cardiac heart disease; ldl - low-density lipoprotein; hdl - high-density lipoprotein; trig - triglycerides and t2d - type 2 diabetes) around the HMGCR (3-hydroxy-3-methylglutaryl-CoA reductase) gene region.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chd_id</span> <span class="op">&lt;-</span> <span class="st">"ieu-a-7"</span></span>
<span><span class="va">ldl_id</span> <span class="op">&lt;-</span> <span class="st">"ieu-b-110"</span></span>
<span><span class="va">hdl_id</span> <span class="op">&lt;-</span> <span class="st">"ieu-b-109"</span></span>
<span><span class="va">trig_id</span> <span class="op">&lt;-</span> <span class="st">"ieu-b-111"</span></span>
<span><span class="va">t2d_id</span> <span class="op">&lt;-</span> <span class="st">"ebi-a-GCST006867"</span></span>
<span><span class="va">hmgcr_chrpos</span> <span class="op">&lt;-</span> <span class="st">"5:74132993-75132993"</span></span>
<span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">chd_id</span>, <span class="va">ldl_id</span>, <span class="va">hdl_id</span>, <span class="va">trig_id</span>, <span class="va">t2d_id</span><span class="op">)</span></span></code></pre></div>
<p>First we will set the path to where the reference population plink files are located to build the LD matrix.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld_ref</span> <span class="op">=</span> <span class="st">"/project/data/reference/ld/EUR"</span></span></code></pre></div>
<p>Create a gwasglue 2 DataSet object and harmonise trait against trait, trait against LD matrix</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create gwasglue2 DataSet object (hmgcr)</span></span>
<span><span class="va">hmgcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createDataSet.html">createDataSet</a></span><span class="op">(</span>traits<span class="op">=</span><span class="va">ids</span>, variants <span class="op">=</span> <span class="va">hmgcr_chrpos</span>, tools <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"finemap"</span>,<span class="st">"coloc"</span><span class="op">)</span>, source <span class="op">=</span> <span class="st">"IEUopenGWAS"</span>, harmonise <span class="op">=</span> <span class="cn">TRUE</span>, tolerance <span class="op">=</span> <span class="fl">0.08</span>, action <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/harmonise_ld.html">harmonise_ld</a></span><span class="op">(</span><span class="va">.</span>, ld_ref <span class="op">=</span> <span class="va">ld_ref</span>, bfile <span class="op">=</span> <span class="cn">TRUE</span>, plink_bin <span class="op">=</span> <span class="st">"plink"</span><span class="op">)</span> </span></code></pre></div>
<p>Before starting with the fine mapping and colocalisation analyses, we are going to look at the p-values for each variant in each trait.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot gwasglue2 DataSet object (hmgcr)</span></span>
<span></span>
<span>  <span class="fu"><a href="../reference/plot_gwasglue.html">plot_gwasglue</a></span><span class="op">(</span><span class="va">hmgcr</span>, type<span class="op">=</span><span class="st">"manhattan"</span><span class="op">)</span></span></code></pre></div>
<!-- After summary sets are harmonised, marginalise each summary set independently and create a new dataset with all marginalised summary sets merged -->
<!-- ```{r} -->
<!-- # do finemapping with susie -->
<!-- ntraits <- getLength(hmgcr) -->
<!-- hmgcr_marginalised <- lapply(1:ntraits, function(trait) -->
<!--   { -->
<!--     # Takes in 1 SS -->
<!--     # Outputs 1 DS (with at least 1 SS) -->
<!--     ds <- susie_rss(R = getLDMatrix(ds_hmgcr,trait), summaryset = getSummarySet(ds_hmgcr, trait)) -->
<!-- })  -->
<!-- ``` -->
<!-- Because not all datasets were marginalised, we are going to merge just the ones that were -->
<!-- ```{r} -->
<!-- hmgcr_marginalised <- merge_datasets(hmgcr_marginalised[[2]], hmgcr_marginalised[[3]]) -->
<!-- ``` -->
<!-- Try hyprcoloc with raw datasets. (Use fork of hypercoloc that allows a dataset object to be provided) -->
<!-- ```{r} -->
<!-- res_hmgcr <- hyprcoloc(dataset = hmgcr) -->
<!-- hyprcoloc is using gwasglue2 DataSet class object as input -->
<!-- print(res_hmgcr) -->
<!-- ``` -->
<!-- Try hyprcoloc with marginalised datasets -->
<!-- ```{r} -->
<!-- res_hmgcr_marginalised <- hyprcoloc(dataset =hmgcr_marginalised) -->
<!-- hyprcoloc is using gwasglue2 DataSet class object as input -->
<!-- print(res_hmgcr_marginalised) -->
<!-- ``` -->
<p>Also try with these other regions:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcsk9_chrpos</span> <span class="op">&lt;-</span> <span class="st">"1:55005221-56005221"</span></span>
<span><span class="va">npc1l1_chrpos</span> <span class="op">&lt;-</span> <span class="st">"7:44052134-45052134"</span></span>
<span><span class="va">lpa_chrpos</span> <span class="op">&lt;-</span> <span class="st">"6:160952515-161087407"</span></span></code></pre></div>
<ul>
<li>PCSK9 (proprotein convertase subtilisin/kexin type 9) gene region</li>
</ul>
<p>Create dataset and harmonise trait against trait, trait against LD matrix (note that we are not saving the SummarySet objects separately, but filling them directly to the DataSet).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create gwasglue2 DataSet object (pcsk9)</span></span>
<span><span class="va">pcsk9</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createDataSet.html">createDataSet</a></span><span class="op">(</span>traits <span class="op">=</span> <span class="va">ids</span>, variants <span class="op">=</span> <span class="va">pcsk9_chrpos</span>, tools <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"finemap"</span>,<span class="st">"coloc"</span><span class="op">)</span>, source <span class="op">=</span> <span class="st">"IEUopenGWAS"</span>, harmonise <span class="op">=</span> <span class="cn">TRUE</span>, tolerance <span class="op">=</span> <span class="fl">0.08</span>, action <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/harmonise_ld.html">harmonise_ld</a></span><span class="op">(</span><span class="va">.</span>, ld_ref <span class="op">=</span> <span class="va">ld_ref</span>, bfile <span class="op">=</span> <span class="cn">TRUE</span>, plink_bin <span class="op">=</span> <span class="st">"plink"</span><span class="op">)</span> </span></code></pre></div>
<p>Manhattan plots for each trait in the DataSet.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot gwasglue2 DataSet object (hmgcr)</span></span>
<span></span>
<span>  <span class="fu"><a href="../reference/plot_gwasglue.html">plot_gwasglue</a></span><span class="op">(</span><span class="va">pcsk9</span>, type<span class="op">=</span><span class="st">"manhattan"</span><span class="op">)</span></span></code></pre></div>
<!-- Marginalise each summary set independently and create a new dataset with all marginalised summary sets merged -->
<!-- ```{r} -->
<!-- # do finemapping with susie -->
<!-- ntraits <- getLength(pcsk9) -->
<!-- pcsk9_marginalised <- lapply(1:ntraits, function(trait) -->
<!--   { -->
<!--     # Takes in 1 SS -->
<!--     # Outputs 1 DS (with at least 1 SS) -->
<!--     ds <- susie_rss(R = getLDMatrix(pcsk9,trait), summaryset = getSummarySet(pcsk9, trait)) -->
<!-- })  -->
<!-- ``` -->
<!-- Because not all datasets were marginalised, we are going to merge just the ones that were -->
<!-- ```{r} -->
<!-- ds_hmgcr_marginalised <- merge_datasets(ds_hmgcr_marginalised[[2]]) -->
<!-- ``` -->
<!-- Try hyprcoloc with raw datasets.  -->
<!-- ```{r} -->
<!-- res_pcsk9 <- hyprcoloc(dataset = pcsk9) -->
<!-- hyprcoloc is using gwasglue2 DataSet class object as input -->
<!-- print(res_pcsk9) -->
<!-- ``` -->
<!-- Try hyprcoloc with marginalised datasets -->
<!-- ```{r} -->
<!-- res_pcsk9_marginalised <- hyprcoloc(dataset =pcsk9_marginalised) -->
<!-- hyprcoloc is using gwasglue2 DataSet class object as input -->
<!-- print(res_pcsk9_marginalised) -->
<!-- $results -->
<!-- ``` -->
<ul>
<li>NPC1L1 (NPC1 like intracellular cholesterol transporter 1) gene region Create dataset and harmonise trait against trait, trait against LD matrix</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create gwasglue2 DataSet object (npc1l1)</span></span>
<span><span class="va">npc1l1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createDataSet.html">createDataSet</a></span><span class="op">(</span>traits<span class="op">=</span><span class="va">ids</span>, variants <span class="op">=</span> <span class="va">npc1l1_chrpos</span>, tools <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"finemap"</span>,<span class="st">"coloc"</span><span class="op">)</span>, source <span class="op">=</span> <span class="st">"IEUopenGWAS"</span>, harmonise <span class="op">=</span> <span class="cn">TRUE</span>, tolerance <span class="op">=</span> <span class="fl">0.08</span>, action <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/harmonise_ld.html">harmonise_ld</a></span><span class="op">(</span><span class="va">.</span>, ld_ref <span class="op">=</span> <span class="va">ld_ref</span>, bfile <span class="op">=</span> <span class="cn">TRUE</span>, plink_bin <span class="op">=</span> <span class="st">"plink"</span><span class="op">)</span> </span></code></pre></div>
<p>Manhattan plots for each trait in the DataSet.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot gwasglue2 DataSet object (hmgcr)</span></span>
<span></span>
<span>  <span class="fu"><a href="../reference/plot_gwasglue.html">plot_gwasglue</a></span><span class="op">(</span><span class="va">npc1l1</span>, type<span class="op">=</span><span class="st">"manhattan"</span><span class="op">)</span></span></code></pre></div>
<!-- Marginalise each summary set independently and create a new dataset with all marginalised summary sets merged -->
<!-- ```{r} -->
<!-- ntraits <- getLength( npc1l1) -->
<!-- npc1l1_marginalised <- lapply(1:ntraits, function(trait) -->
<!--   { -->
<!--     # Takes in 1 SS -->
<!--     # Outputs 1 DS (with at least 1 SS) -->
<!--     ds <- susie_rss(R = getLDMatrix(npc1l1,trait), summaryset = getSummarySet( npc1l1, trait)) -->
<!-- })  -->
<!-- ``` -->
<!-- Because not all datasets were marginalised, we are going to merge just the ones that were -->
<!-- ```{r} -->
<!-- npc1l1_marginalised <- merge_datasets(npc1l1_marginalised[[2]],npc1l1_marginalised[[3]],npc1l1_marginalised[[4]],npc1l1_marginalised[[5]]) -->
<!-- ``` -->
<!-- Try hyprcoloc with raw datasets.  -->
<!-- ```{r} -->
<!-- res_npc1l1 <- hyprcoloc(dataset = npc1l1) -->
<!-- hyprcoloc is using gwasglue2 DataSet class object as input -->
<!-- print(res_npc1l1) -->
<!-- $results -->
<!-- ``` -->
<!-- Try hyprcoloc with marginalised datasets -->
<!-- ```{r} -->
<!-- res_npc1l1_marginalised <- hyprcoloc(dataset =npc1l1_marginalised) -->
<!-- hyprcoloc is using gwasglue2 DataSet class object as input -->
<!-- print(res_npc1l1_marginalised) -->
<!-- ``` -->
<ul>
<li>LPA (Lipoprotein(A)) gene region Create dataset and harmonise trait against trait, trait against LD matrix</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create gwasglue2 DataSet object (lpa)</span></span>
<span><span class="va">lpa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createDataSet.html">createDataSet</a></span><span class="op">(</span>traits<span class="op">=</span><span class="va">ids</span>, variants <span class="op">=</span> <span class="va">lpa_chrpos</span>, tools <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"finemap"</span>,<span class="st">"coloc"</span><span class="op">)</span>, source <span class="op">=</span> <span class="st">"IEUopenGWAS"</span>, harmonise <span class="op">=</span> <span class="cn">TRUE</span>, tolerance <span class="op">=</span> <span class="fl">0.08</span>, action <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/harmonise_ld.html">harmonise_ld</a></span><span class="op">(</span><span class="va">.</span>, ld_ref <span class="op">=</span> <span class="va">ld_ref</span>, bfile <span class="op">=</span> <span class="cn">TRUE</span>, plink_bin <span class="op">=</span> <span class="st">"plink"</span><span class="op">)</span> </span></code></pre></div>
<p>Manhattan plots for each trait in the DataSet.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot gwasglue2 DataSet object (hmgcr)</span></span>
<span></span>
<span>  <span class="fu"><a href="../reference/plot_gwasglue.html">plot_gwasglue</a></span><span class="op">(</span><span class="va">lpa</span>, type<span class="op">=</span><span class="st">"manhattan"</span><span class="op">)</span></span></code></pre></div>
<!-- Marginalise each summary set independently and create a new dataset with all marginalised summary sets merged -->
<!-- ```{r} -->
<!-- ntraits <- getLength(lpa) -->
<!-- npc1l1_marginalised <- lapply(1:ntraits, function(trait) -->
<!-- { -->
<!--   # Takes in 1 SS -->
<!--   # Outputs 1 DS (with at least 1 SS) -->
<!--   ds <- susie_rss(R = getLDMatrix(lpa,trait), summaryset = getSummarySet(lpa, trait)) -->
<!-- })  -->
<!-- lpa_marginalised <- merge_datasets() -->
<!-- ``` -->
<!-- Because not all datasets were marginalised, we are going to merge just the ones that were -->
<!-- ```{r} -->
<!-- ``` -->
<!-- Try hyprcoloc with raw datasets.  -->
<!-- ```{r} -->
<!-- res_lpa <- hyprcoloc(dataset = nlpa) -->
<!-- print(res_lpa) -->
<!-- ``` -->
<!-- Try hyprcoloc with marginalised datasets -->
<!-- ```{r} -->
<!-- res_lpa_marginalised <- hyprcoloc(dataset =lpa_marginalised) -->
<!-- print(res_lpa_marginalised) -->
<!-- ``` -->
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Gibran Hemani, Rita Rasteiro, Christopher Woods.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
